enclave/core/sgx/exception.c

```diff
@@ -456,29 +462,36 @@ void oe_real_exception_dispatcher(oe_context_t* oe_context)
         oe_abort();
         return;
     }
-    else if (handler_ret == OE_EXCEPTION_ABORT_EXECUTION)
+
+    if (is_segfault)
+        OE_TRACE_ERROR("Segmentation fault");
+    else
+        OE_TRACE_ERROR("Unhandled exception");
+
+    void* buffer[OE_BACKTRACE_MAX];
+    oe_result_t r = OE_UNEXPECTED;
+    int size;
+
+    /* Get the backtrace from the previous frame where the exception occurs */
+    if ((size = _backtrace_skip_current_frame(buffer, OE_BACKTRACE_MAX)) > 0)
     {
-        oe_abort();
-        return;
+        oe_sgx_log_backtrace_ocall(
+            &r,
+            oe_get_enclave(),
+            OE_LOG_LEVEL_ERROR,
+            (uint64_t*)buffer,
+            (size_t)size);
     }
 
     // Exception can't be handled by trusted handlers, abort the enclave.
-    // Let the oe_abort to run on the stack where the exception happens.
-    // oe_continue_execution uses a retq instruction to jump to the target
-    // address. Since no return address is being pushed to the stack, this
-    // can break x64 ABI compatibility which requires that RSP ought to
-    // be a multiple of 8 upon entry to the function. To ensure compatibility,
-    // first align the RSP to 16 bytes and the push the current rsp to fake
-    // a call.
-    oe_exception_record.context->rsp =
-        (oe_exception_record.context->rsp & (uint64_t)-16) - 8;
-    uint64_t current_rip = 0;
-    asm volatile("lea (%%rip), %0" : "=r"(current_rip));
-    *(uint64_t*)oe_exception_record.context->rsp = current_rip;
-    oe_exception_record.context->rip = (uint64_t)oe_abort;
-    oe_continue_execution(oe_exception_record.context);
+    oe_abort();
 
-    return;
+    // Code should never run to here.
+}
+
+static int _backtrace_skip_current_frame(void** buffer, int size)
+{
+    return oe_backtrace_impl(__builtin_frame_address(1), buffer, size);
```

Comment:
This may work differently depending on whether _backtrac_skip_current_frame is inlined or not.